README.md
# NAUIViewWithBorders

UIView subclass that enables selective borders. (e.g. only a bottom border)  
Each side can be figured with a different color and width.

![alt tag](http://natrosoft.com/wp-content/uploads/2014/01/oneBorder.png)

![alt tag](http://natrosoft.com/wp-content/uploads/2014/01/differentBorders.png)


## Installation
NAUIViewWithBorders is available through [CocoaPods](http://cocoapods.org), to install
it simply add the following line to your Podfile:  
    pod "NAUIViewWithBorders"

## Author
Nathan Rowe, natrosoft@gmail.com

## License
MIT license. See the LICENSE file for more info.

## Usage
For an example, open the demo workspace in Example folder.

**Basic use case:**
- Change your UIView in Interface Builder to class type NAUIViewWithBorders
- In your view controller's viewDidLoad
```
    /* For a top border ---------------------------------------------- */
    self.myBorderView.borderColorTop = [UIColor redColor];
    self.myBorderView..borderWidthsAll  = 1.0f;

    /* For borders with different colors and widths --------------------------- */
    self.myBorderView.borderWidths      = UIEdgeInsetsMake(2.0, 4.0, 6.0, 8.0);
    self.myBorderView.borderColorTop    = [UIColor blueColor];
    self.myBorderView.borderColorRight  = [UIColor redColor];
    self.myBorderView.borderColorBottom = [UIColor greenColor];
    self.myBorderView.borderColorLeft   = [UIColor darkGrayColor];

    /* For changing the draw order so that the top and bottom borders are draw on top like so ----------------------------
     *    ---------
     *    |       |
     *    ---------  
     */
    self.myBorderView.borderWidthsAll     = 8.0;
    self.myBorderView.borderColorTop      = [UIColor blueColor];
    self.myBorderView.borderColorBottom   = [UIColor blueColor];
    self.myBorderView.borderColorLeft     = [UIColor greenColor];
    self.myBorderView.borderColorRight    = [UIColor greenColor];
    self.myBorderView..drawOrder = [NSOrderedSet orderedSetWithObjects:kNABorderLeft, kNABorderRight,
                                                                       kNABorderTop, kNABorderBottom, nil];

    /* or with side borders drawn on top
     *    |---------|
     *    |         |
     *    |---------|
     */
    self.myBorderView..drawOrder = [NSOrderedSet orderedSetWithObjects:kNABorderTop, kNABorderBottom,
                                                                       kNABorderLeft, kNABorderRight, nil];
```


Notes:
- Borders are always drawn from the view's bounds inward. (border line will not spill outside view)
- Don't create a similar subclass to UIImageView (e.g. UIImageViewWithBorders) to get selective borders on a UIImageView.
  UIImageView does not call drawRect of its sublcasses, so the code to draw the selective borders will not run.  Place
  an NAUIViewWithBorders over top of a UIImageView to achieve selective borders.
- Be sure to check out the demo project.


 
![alt tag](http://natrosoft.com/wp-content/uploads/2014/01/selectiveBorderSS.jpg)
 

[![githalytics.com alpha](https://cruel-carlota.pagodabox.com/1b4eef099ab4ddc7be7d89fc151c3dd8 "githalytics.com")](http://githalytics.com/natrosoft/NAUIViewWithBorders)
    
    
 


CHANGELOG.md
# NAUIViewWithBorders CHANGELOG

## 0.1.0

Initial release.


Example/NAUIViewWithBordersDemo/NAViewController.h
//
//  NAViewController.h
//  NAUIViewWithBordersDemo
//
//  Created by Nathan Rowe on 1/10/14.
//  Copyright (c) 2014 Natrosoft LLC. All rights reserved.
//

#import <UIKit/UIKit.h>
#import "NAUIViewWithBorders.h"

@interface NAViewController : UIViewController
{}
@property (nonatomic, weak) IBOutlet NAUIViewWithBorders *simpleExampleView;
@property (nonatomic, weak) IBOutlet NAUIViewWithBorders *differentBordersView;
@property (nonatomic, weak) IBOutlet NAUIViewWithBorders *differentDrawOrderView;
@property (nonatomic, weak) IBOutlet UILabel *differentDrawOrderLabel;
@property (nonatomic, weak) IBOutlet NAUIViewWithBorders *imageOverlayView;
@property (nonatomic, weak) IBOutlet NAUIViewWithBorders *bordersDrawnWithinLeftView;
@property (nonatomic, weak) IBOutlet NAUIViewWithBorders *bordersDrawnWithinRightView;


@end


NAUIViewWithBordersDemo/NAUIViewWithBordersDemo/NAAppDelegate.m
//
//  NAAppDelegate.m
//  NAUIViewWithBordersDemo
//
//  Created by Nathan Rowe on 1/10/14.
//  Copyright (c) 2014 Natrosoft LLC. All rights reserved.
//

#import "NAAppDelegate.h"

@implementation NAAppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    // Override point for customization after application launch.
    return YES;
}
							
- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later. 
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


NAUIViewWithBordersDemo/NAUIViewWithBordersDemo/NAViewController.m
//
//  NAViewController.m
//  NAUIViewWithBordersDemo
//
//  Created by Nathan Rowe on 1/10/14.
//  Copyright (c) 2014 Natrosoft LLC. All rights reserved.
//

#import "NAViewController.h"

@interface NAViewController ()
typedef void (^GCDSafeTimer)(void);
@property (nonatomic, copy) GCDSafeTimer safeTimer;
@end



#pragma mark -
@implementation NAViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    /* Simple example with only a top border
     * - simply do not set the other border colors and they will not be drawn */
    self.simpleExampleView.borderColorTop   = [UIColor redColor];
    self.simpleExampleView.borderWidthsAll  = 1.0f;

    /* Example of each side of a view having different border colors and widths */
    self.differentBordersView.borderWidths      = UIEdgeInsetsMake(2.0, 4.0, 6.0, 8.0);
    self.differentBordersView.borderColorTop    = [UIColor blueColor];
    self.differentBordersView.borderColorRight  = [UIColor redColor];
    self.differentBordersView.borderColorBottom = [UIColor greenColor];
    self.differentBordersView.borderColorLeft   = [UIColor darkGrayColor];
    
    /* Example to illustrate how the draw order matters.  The default is clockwise
     * (top, right, bottom, left) as can be seen by how the lines overlap on the
     * corners with different colors.  The draw order can be changed. */
    self.differentDrawOrderView.borderWidthsAll     = 8.0;
    self.differentDrawOrderView.borderColorTop      = [UIColor blueColor];
    self.differentDrawOrderView.borderColorBottom   = [UIColor redColor];
    self.differentDrawOrderView.borderColorLeft     = [UIColor greenColor];
    self.differentDrawOrderView.borderColorRight    = [UIColor darkGrayColor];

    //Just set the draw order like so:
    self.differentDrawOrderView.drawOrder = [NSOrderedSet orderedSetWithObjects:kNABorderTop, kNABorderRight, kNABorderBottom, kNABorderLeft, nil];

    //Animation to illustrate what different draw orders look like
    [self startDrawOrderExampleAnimation];
    
    
    /* Example for drawing selective borders on image views.
     * NAUIImageWithBorders is a simple UIView subclass, so simply place it over top
     * of a UIImageView to simulate a UIImageView with selective borders. */
    self.imageOverlayView.borderWidthsAll = 2.0;
    self.imageOverlayView.borderColorRight = [UIColor blackColor];
    self.imageOverlayView.borderColorLeft = [UIColor blackColor];
    

    /* Example for knowing how to align borders.  Borders are drawn from the view's edge inward.
       They do not "spill" outside of its bounds */
    self.bordersDrawnWithinLeftView.borderWidthsAll = 3.0;
    self.bordersDrawnWithinLeftView.borderColorRight = [UIColor lightGrayColor];
    self.bordersDrawnWithinLeftView.borderColorBottom = [UIColor lightGrayColor];
    self.bordersDrawnWithinRightView.borderWidthsAll = 1.0;
    self.bordersDrawnWithinRightView.borderColorLeft = [UIColor blackColor];
    self.bordersDrawnWithinRightView.borderColorBottom = [UIColor blackColor];
    
}



- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}





//---------------------------
#pragma mark - Helper Methods
//---------------------------
- (void) startDrawOrderExampleAnimation
{
    //Create different draw orders
    NSOrderedSet *orderOne      = [NSOrderedSet orderedSetWithObjects:kNABorderTop, kNABorderRight, kNABorderBottom, kNABorderLeft, nil];
    NSString *orderOneLabel     = @"Clockwise draw order";
    self.differentDrawOrderLabel.text = orderOneLabel;
    NSOrderedSet *orderTwo      = [NSOrderedSet orderedSetWithObjects:kNABorderTop, kNABorderLeft, kNABorderBottom, kNABorderRight, nil];
    NSString *orderTwoLabel     = @"Counter-Clockwise draw order";
    NSOrderedSet *orderThree    = [NSOrderedSet orderedSetWithObjects:kNABorderTop, kNABorderBottom, kNABorderLeft, kNABorderRight, nil];
    NSString *orderThreeLabel   = @"Side borders drawn last";
    NSOrderedSet *orderFour     = [NSOrderedSet orderedSetWithObjects:kNABorderLeft, kNABorderRight, kNABorderTop, kNABorderBottom, nil];
    NSString *orderFourLabel    = @"Top & bottom borders drawn last";
    
    //Change draw order and force a redraw
    __weak id weakself = self;
    self.safeTimer = ^{
        double delaySeconds = 3.5;
        dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, delaySeconds * NSEC_PER_SEC);
        dispatch_after(popTime, dispatch_get_main_queue(), ^(void) {
            NAViewController *strongself = weakself;
            if (strongself)
            {
                //Cycle through different draw orders
                if (!strongself.differentDrawOrderView.drawOrder || [strongself.differentDrawOrderView.drawOrder isEqualToOrderedSet:orderFour]) {
                    strongself.differentDrawOrderView.drawOrder = orderOne;
                    strongself.differentDrawOrderLabel.text = orderOneLabel;
                }
                else if ([strongself.differentDrawOrderView.drawOrder isEqualToOrderedSet:orderOne]) {
                    strongself.differentDrawOrderView.drawOrder = orderTwo;
                    strongself.differentDrawOrderLabel.text = orderTwoLabel;
                }
                else if ([strongself.differentDrawOrderView.drawOrder isEqualToOrderedSet:orderTwo]) {
                    strongself.differentDrawOrderView.drawOrder = orderThree;
                    strongself.differentDrawOrderLabel.text = orderThreeLabel;
                }
                else if ([strongself.differentDrawOrderView.drawOrder isEqualToOrderedSet:orderThree]) {
                    strongself.differentDrawOrderView.drawOrder = orderFour;
                    strongself.differentDrawOrderLabel.text = orderFourLabel;
                }
                
                [strongself.differentDrawOrderView setNeedsDisplay];    /* redraw */
                
                //Run again
                strongself.safeTimer();
            }
        });
    };
    self.safeTimer();   //start running
}


@end
